### PR TITLE
Integrate with Fetch Metadata.

### LINKED ISSUE #885: Consider supporting Fetch Metadata.
https://mikewest.github.io/sec-metadata/ is starting to look pretty reasonable. I plan to put up some PRs to sketch out what the integration would look like, based on the thoughts in https://mikewest.github.io/sec-metadata/#fetch-integration. I think we'd need ~3 changes, but who knows? :)

1.  Add a boolean `user activation` flag to `request` objects, defaulting to `false`. We'd set it in HTML's navigation algorithm, based on whether the navigation was triggered by user activation.

2.  Call [set the Fetch Metadata headers for a request](https://mikewest.github.io/sec-metadata/#abstract-opdef-set-the-fetch-metadata-headers-for-a-request). Probably from "main fetch"?

3.  Define "`nested-navigate`", as per #755.

WDYT?

### PR DESCRIPTION
**Pull request by mikewest**
This patch integrates Fetch Metadata processing into Fetch's "main
fetch" algorithm, and defines a "user activation flag" on requests that
will be populated during HTML's "process a navigate fetch" algorithm.

Closes #885.

- [X] At least two implementers are interested (and none opposed):
    * Chrome is shipping this as of 76.
    * Mozilla has labeled the feature as "[worth prototyping](https://mozilla.github.io/standards-positions/#fetch-metadata)".
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
    * https://wpt.fyi/results/fetch/metadata?label=experimental&label=master&aligned
    * @mikewest is writing a test for the specific bit around Service Worker's access to the headers in https://github.com/web-platform-tests/wpt/pull/27857.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
    * Chrome: Shipped.
    * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1695911
    * Safari: https://bugs.webkit.org/show_bug.cgi?id=204744

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/fetch/993.html" title="Last updated on Mar 4, 2021, 8:14 AM UTC (b36a125)">Preview</a> | <a href="https://whatpr.org/fetch/993/24a8670...b36a125.html" title="Last updated on Mar 4, 2021, 8:14 AM UTC (b36a125)">Diff</a>

### CODE REVIEW

**Comment by mikewest** [comment]
As discussed on public-webappsec@. WDYT, @annevk?

File: fetch.bs
Comment (annevk) [inline fetch.bs]: This should be an exported term. You can file a bug against tabatkins/bikeshed to get Fetch Metadata indexed. If all goes well we can land this PR Monday.

Comment (mikewest) [reply fetch.bs]: Hrm. It is exported from Fetch Metadata, and that spec shows up in Bikeshed's biblio entries, but not in its anchor index. Weird!

Filed https://github.com/tabatkins/bikeshed/issues/1584

Comment (annevk) [reply fetch.bs]: Biblio is for references at the end of the document, Shepherd is for inline cross-references and does not pick up new specifications to crawl automatically.

**Comment by mikewest** [comment]
Coming back to this as a prereq to https://github.com/whatwg/html/pull/5203.

WDYT?

File: fetch.bs
Comment (annevk) [inline fetch.bs]: Which is this an abstract-op? I thought those were only for JavaScript.

Comment (mikewest) [reply fetch.bs]: Bikeshed defines `abstract-op` as "[English-language algorithms](https://tabatkins.github.io/bikeshed/#dfn-types)"; that's generally why I've been using it to identify algorithms. What type would you use?

**Comment by mikewest** [comment]
> As Yoav once discovered, this does not work. As https://fetch.spec.whatwg.org/#cors-unsafe-request-header-names is currently invoked, setting Sec- headers before the network layer forces a preflight. We'd have to add an exception there (or its caller).

Would you like me to put these headers specifically into https://fetch.spec.whatwg.org/#cors-safelisted-request-header? Or would you like us to work out a `Sec-*` carveout generally?

File: fetch.bs
Comment (annevk) [reply fetch.bs]: Interesting, I tend to use dfn for those and only use abstract-op for the IAmACamelCaseAlgorithm types from ECMAScript. I guess I'm not going to block on this then, but copying @domenic and @tabatkins for thoughts.

**Comment by annevk** [comment]
I think we should carve it out generically, right?

Now actually copying @yoavweiss so he can share any thoughts as to why carving out `Sec-` generically might be bad.

**Comment by mikewest** [comment]
Would you like me to carve it out generically in this CL, or in a separate patch? Probably the latter, right?

**Comment by annevk** [comment]
This also does not address https://github.com/w3c/webappsec-fetch-metadata/issues/29 as far as I can tell. We need a solution for that.

Also, https://github.com/w3c/webappsec-fetch-metadata/issues/38 changed right as this PR would expose them to service workers?

Could you maybe do another triage round so I know what upstream issues might end up impacting Fetch still?

File: fetch.bs
Comment (tabatkins) [reply fetch.bs]: I think it's appropriate to use abstract-op here, fwiw.

Comment (tabatkins) [reply fetch.bs]: (When I finally add some cool functionality around registering algorithms with arguments and types and whatnot, I expect that I'll hook abstract-op as the type for that.)

**Comment by mikewest** [comment]
I merged the last few months of changes in https://github.com/whatwg/fetch/pull/993/commits/f7800972676ffd16648ed401603ce3a9696a6fbe, and moved the integration out of "main fetch" and into "HTTP-network-or-cache" fetch in https://github.com/whatwg/fetch/pull/993/commits/58d9d1870d9adc2bd348033c3bf897fe38729ae2. I believe that addresses the following concerns:

*   We do not expose the headers to the `Request.headers` accessor in Service workers, as they're added after SW have a shot at handling the request. That should allow us to resolve https://github.com/w3c/webappsec-fetch-metadata/issues/38.
*   The HTTPS->HTTP->HTTPS redirect behavior described in the spec (and in WPT) should fall naturally out of this change for the same reason (we recalculate on each hop), which should allow us to resolve https://github.com/w3c/webappsec-fetch-metadata/issues/29.
*    We can punt the question of a more generic `Sec-*` carveout for CORS preflights, as this change ensures that headers are added after the point at which we would have send a preflight (in HTTP fetch).

I've skimmed very quickly through other bugs, but nothing other than https://github.com/whatwg/html/pull/5203 jumps out at me as needing to be resolved before landing this integration. I'll try to label things accordingly if I can find some time.

WDYT?

File: fetch.bs
Comment (annevk) [inline fetch.bs:1667]: Let's change this into a boolean, similar to keepalive. I'd like to avoid adding flags and slowly convert existing flags to booleans. (I'm happy to help with this for the Fetch and HTML PRs if needed, but would prefer if someone else did it for Fetch Metadata.)

I think we should also add a note that this is for exclusive usage by HTML's navigate algorithm. At least for now?

File: fetch.bs
Comment (annevk) [inline fetch.bs:4212]: _httpRequest_, right?

Comment (annevk) [reply fetch.bs:4212]: And ideally we would use "append" here as we do elsewhere since there should be nothing to override.

**Review by annevk — COMMENTED** [review]
Thanks for getting around to updating this. The plan looks good to me, I only have nits.

**Comment by annevk** [comment]
I've also added the pull request template to OP as this is a normative change. Mainly so we can find tests and bugs later. I think this has support from Chrome/Firefox.

Maybe @youennf can weigh in for Safari?

File: fetch.bs
Comment (mikewest) [reply fetch.bs]: Updated in https://github.com/whatwg/fetch/pull/993/commits/c0050629590f01932fcaf4931699705c54029daa. I'll take care of the HTML PR and Fetch Metadata once you're happy with this patch.

File: fetch.bs
Comment (mikewest) [reply fetch.bs]: Used the right variable, and renamed the algorithm. I'll update Fetch Metadata accordingly once you're happy with the rest.

**Review by annevk — APPROVED** [review]
Yeah this looks good, thanks.

File: fetch.bs
Comment (mikewest) [reply fetch.bs]: HTML is addressed in the updates to https://github.com/whatwg/html/pull/5203.

Fetch Metadata is renamed as well: https://w3c.github.io/webappsec-fetch-metadata/#abstract-opdef-append-the-fetch-metadata-headers-for-a-request.

**Comment by mikewest** [comment]
Updated dependent PRs and specs, and filled out the template above.

**Comment by annevk** [comment]
Thanks @mikewest!

**Comment by annevk** [comment]
Unfortunately, I just realized I forgot to note https://github.com/whatwg/html/pull/5203 in the commit message. Oh well.

**Pull request merged**
